feat: insights routing forms

sean-brydon opened the pull request:
## What does this PR do?

Fixes: CAL-4615

loom: https://cap.so/s/m4jhnp9ayj3ranp
## Mandatory Tasks (DO NOT REMOVE)

- [ ] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [ ] I have updated the developer docs in /docs if this PR makes changes that would require a [documentation change](https://cal.com/docs). If N/A, write N/A here and check the checkbox.
- [ ] I confirm automated tests are in place that prove my fix is effective or that my feature works.

## How should this be tested?
1) seed org with new insights seed script.
2) View insights/routing
3) Try filters.
4) Enjoy

## Checklist

<!-- Remove bullet points below that don't apply to you -->

- I haven't read the [contributing guide](https://github.com/calcom/cal.com/blob/main/CONTRIBUTING.md)
- My code doesn't follow the style guidelines of this project
- I haven't commented my code, particularly in hard-to-understand areas
- I haven't checked if my changes generate no new warnings

sean-brydon committed:
seed to org

sean-brydon committed:
more org+routing form seed improvments

sean-brydon committed:
add routing responses

sean-brydon committed:
check if attributes already exist to prevent crash

sean-brydon committed:
remove files not for PR

sean-brydon committed:
fix type error

sean-brydon committed:
fix connection

sean-brydon committed:
assign manually

sean-brydon committed:
Merge branch 'main' into feat/routingform-org-attribute-seed

sean-brydon committed:
fix seed

sean-brydon committed:
make schemas reusable

sean-brydon committed:
add Sub route for routing specific insights

sean-brydon committed:
add i18n for nav item

sean-brydon committed:
add routing forms new page + layout + basig routing insights class

sean-brydon committed:
basic filters per routing form

sean-brydon committed:
Merge branch 'main' into feat/routingform-org-attribute-seed

sean-brydon committed:
Merge branch 'main' into feat/insights-routing-forms

sean-brydon committed:
use same design as data-table for filters

sean-brydon committed:
add new filter component to eventtypelist

sean-brydon committed:
pass routingFormId to query

sean-brydon committed:
fix types

sean-brydon committed:
Merge branch 'main' into feat/routingform-org-attribute-seed

sean-brydon committed:
add infinite table for responses + hover card on user

sean-brydon committed:
fix: working people query

sean-brydon committed:
working kpi filter with people filter

sean-brydon committed:
add form responses to header

sean-brydon committed:
feat/working-dynamic-field values

sean-brydon committed:
add hover card to skills instead of tooltip

sean-brydon committed:
i18n + allign no booking badge with placeholder avatar

sean-brydon committed:
improve hovercard+start on attribute filters

sean-brydon committed:
add i18n

sean-brydon committed:
fix badge status

sean-brydon committed:
booking status filters - dynamic attirbute filters

sean-brydon committed:
working serverside booking status filters

sean-brydon committed:
working dynamic filters

sean-brydon committed:
add routing responses filter to KPI cards

sean-brydon committed:
adds booking attendees to table

sean-brydon committed:
move columns around

sean-brydon committed:
wip SQL Logic

sean-brydon committed:
group fields

zomars reviewed (commented):
> Loom: https://cap.link/gsqgvhbm07q0ee4

I can't see the recording. It just stays loading

sean-brydon committed:
create responses without bookings

sean-brydon committed:
update to match integration tests

sean-brydon committed:
show overflow indicator on emails

sean-brydon committed:
move headers to single api call

sean-brydon committed:
extract to components to provide ID

sean-brydon committed:
remove submittedAt

sean-brydon committed:
use most popular as default

sean-brydon committed:
cleanup

sean-brydon committed:
i18n

sean-brydon committed:
fix hover card

sean-brydon committed:
fixed width on cells

sean-brydon committed:
use mostPopular as default

sean-brydon committed:
fix integration test to new user count

sean-brydon commented:
> > Loom: https://cap.link/gsqgvhbm07q0ee4
> 
> I can't see the recording. It just stays loading

Updated

sean-brydon commented on apps/web/modules/insights/insights-view.tsx:
Moving "layout" to a seprate file

sean-brydon committed:
remove overflow-hidden

sean-brydon commented on packages/features/insights/components/RoutingFormResponsesTable.tsx:
```diff
@@ -0,0 +1,371 @@
+        routedToBooking: response.routedToBooking,
+      };
+
+      Object.entries(response.response).forEach(([fieldId, field]) => {
+        const header = headers?.find((h) => h.id === fieldId);
+
+        if (header?.options) {
+          if (Array.isArray(field.value)) {
+            // Map the IDs to their corresponding labels for array values
+            const labels = field.value.map((id) => {
+              const option = header.options?.find((opt) => opt.id === id);
+              return option?.label ?? id;
+            });
+            row[fieldId] = labels;
+          } else {
+            // Handle single value case
+            const option = header.options?.find((opt) => opt.id === field.value);
+            row[fieldId] = option?.label ?? field.value;
+          }
+        } else {
+          row[fieldId] = field.value;
+        }
+      });
```
Get the field values of the forms that we already fetched from headers otherwise the values inside the form are UUIDs  as thats how they are stored in JSON

sean-brydon commented on packages/features/insights/context/FiltersProvider.tsx:
```diff
@@ -16,9 +18,14 @@ const querySchema = z.object({
   memberUserId: z.coerce.number().nullable(),
   eventTypeId: z.coerce.number().nullable(),
-  filter: z.enum(["event-type", "user"]).nullable(),
+  filter: z
```
Add routing_form and bookingstatus as valid options 

Also allow rf_* for any dynamic fields that come from the routing form specifcially

sean-brydon commented on packages/features/insights/context/provider.ts:
```diff
@@ -1,16 +1,23 @@
   selectedEventTypeId?: number | null;
+  selectedRoutingFormId?: string | null;
+  selectedBookingStatus?: BookingStatus | "NO_BOOKING" | null;
+  selectedRoutingFormFilter?: {
+    fieldId: string;
+    optionId: string;
```
We need to store fieldID and the optionId for the backend logic

sean-brydon commented on packages/features/insights/filters/EventTypeList.tsx:
```diff
@@ -29,52 +28,19 @@ export const EventTypeList = memo(() => {
+  const filterOptions = data.map(mapEventTypeToOption);
 
   return (
-    <AnimatedPopover text={getPopoverText()}>
```
Matched designs of data-table filters

sean-brydon commented on packages/features/insights/filters/RoutingFormFieldFilter.tsx:
```diff
@@ -0,0 +1,69 @@
+      label: option.label,
+    })) ?? [];
+
+  console.log("selectedRoutingFormFilter", selectedRoutingFormFilter);
```
```suggestion
```

sean-brydon commented on packages/features/insights/filters/index.tsx:
```diff
@@ -44,7 +47,15 @@ export const Filters = () => {
         <EventTypeList />
 
-        <FilterType />
+        {showRoutingFilters ? (
```
Only show these additional filters on routingForms

sean-brydon commented on packages/features/insights/hooks/useDefaultRoutingForm.ts:
```diff
@@ -0,0 +1,60 @@
+    : null;
+
+  // Effect to set default routing form only once when the page loads
+  useEffect(() => {
+    if (isRoutingInsights && !routingFormId && mostPopularForm && !hasSetDefault.current) {
+      hasSetDefault.current = true;
+      const newSearchParams = new URLSearchParams(searchParams?.toString() ?? undefined);
+      newSearchParams.set("routingFormId", mostPopularForm.id);
+      router.push(`${pathname}?${newSearchParams.toString()}`);
+      onRoutingFormChange(mostPopularForm.id);
+    }
+  }, [
+    isRoutingInsights,
+    routingFormId,
+    mostPopularForm,
+    pathname,
+    searchParams,
+    router,
+    onRoutingFormChange,
+  ]);
```
On inital mount we set the selectedRoutingFormId to the most popular one

sean-brydon replied:
This data is already in react query cache so just re-calling the hook here has no effect

sean-brydon commented on packages/features/insights/server/routing-events.ts:
```diff
@@ -0,0 +1,459 @@
+};
+
+class RoutingEventsInsights {
+  private static async getWhereForTeamOrAllTeams({
```
Reuseable logic used for getting the logic within teams

sean-brydon commented on packages/features/insights/server/routing-events.ts:
```diff
@@ -0,0 +1,459 @@
+    });
+
+    // Base where condition for responses
+    const responsesWhereCondition: Prisma.App_RoutingForms_FormResponseWhereInput = {
```
Combine all active filters

sean-brydon commented on packages/features/insights/server/routing-events.ts:
```diff
@@ -0,0 +1,459 @@
+    };
+  }
+
+  static async getRoutingFormsForFilters({
```
Gets all the routing forms just for the "filters" section

sean-brydon commented on packages/features/insights/server/routing-events.ts:
```diff
@@ -0,0 +1,459 @@
+      where: responsesWhereCondition,
+    });
+
+    const responsesPromise = prisma.app_RoutingForms_FormResponse.findMany({
```
Looks like a lot but this data is pretty small when fetched in chunks of 10 at a time.

sean-brydon commented on packages/features/insights/server/routing-events.ts:
```diff
@@ -0,0 +1,459 @@
+    };
+  }
+
+  static async getRoutingFormFieldOptions({
```
Loads all the options for each "field"

sean-brydon commented on packages/features/insights/server/routing-events.ts:
```diff
@@ -0,0 +1,459 @@
+
+    const teamConditions = [];
+
+    // @ts-expect-error it doest exist but TS isnt smart enough when its unmber or int filter
```
This is where it gets interesting - constructing raw SQL for teams conditions based on the above logic

joeauyeung replied:
```suggestion
    // @ts-expect-error it doest exist but TS isnt smart enough when its number or int filter
```

sean-brydon commented on packages/features/insights/server/routing-events.ts:
```diff
@@ -0,0 +1,459 @@
+        count: number;
+      }[]
+    >`
+      WITH form_fields AS (
+        SELECT 
+          f.id as form_id,
+          f.name as form_name,
+          field->>'id' as field_id,
+          field->>'label' as field_label,
+          opt->>'id' as option_id,
+          opt->>'label' as option_label
+        FROM "App_RoutingForms_Form" f,
+        LATERAL jsonb_array_elements(f.fields) as field
+        LEFT JOIN LATERAL jsonb_array_elements(field->'options') as opt ON true
+        WHERE true
+        ${whereClause}
+      ),
```
Unpacks JSON data from routing form table into a CTE called form fields. 
Also unpacks options of that json into a "id" and "label" so we can use it down the line

sean-brydon commented on packages/features/insights/server/routing-events.ts:
```diff
@@ -0,0 +1,459 @@
+        WHERE true
+        ${whereClause}
+      ),
+      response_stats AS (
+        SELECT 
+          r."formId",
+          key as field_id,
+          CASE 
+            WHEN jsonb_typeof(value->'value') = 'array' THEN
+              v.value_item
+            ELSE
+              value->>'value'
+          END as selected_option,
+          COUNT(DISTINCT r.id) as response_count
+        FROM "App_RoutingForms_FormResponse" r
+        CROSS JOIN jsonb_each(r.response::jsonb) as fields(key, value)
+        LEFT JOIN LATERAL jsonb_array_elements_text(
+          CASE 
+            WHEN jsonb_typeof(value->'value') = 'array' 
+            THEN value->'value'
+            ELSE NULL
+          END
+        ) as v(value_item) ON true
+        WHERE r."routedToBookingUid" IS NULL
+        GROUP BY r."formId", key, selected_option
+      )
```
Counts responses from each form/field/option combination. Excludes responses that were routed to a booking as we only care about "failed" responses (routing responses that have no booking attached"

sean-brydon commented on packages/features/insights/server/routing-events.ts:
```diff
@@ -0,0 +1,459 @@
+        WHERE r."routedToBookingUid" IS NULL
+        GROUP BY r."formId", key, selected_option
+      )
+      SELECT 
+        ff.form_id as "formId",
+        ff.form_name as "formName",
+        ff.field_id as "fieldId",
+        ff.field_label as "fieldLabel",
+        ff.option_id as "optionId",
+        ff.option_label as "optionLabel",
+        COALESCE(rs.response_count, 0)::integer as count
+      FROM form_fields ff
+      LEFT JOIN response_stats rs ON 
+        rs."formId" = ff.form_id AND 
+        rs.field_id = ff.field_id AND 
+        rs.selected_option = ff.option_id
+      WHERE ff.option_id IS NOT NULL
+      ORDER BY count DESC`;
```
Finally select all the key infomation from the CTEs to provide a nice way to see what forms+fields have the most amount of failed bookings booked by form+fieldID

sean-brydon commented on packages/prisma/schema.prisma:
```diff
@@ -603,6 +603,7 @@ model Booking {
   createdAt                    DateTime                       @default(now())
   updatedAt                    DateTime?
   status                       BookingStatus                  @default(ACCEPTED)
+  okin
```
```suggestion
```

sean-brydon commented on packages/prisma/seed-insights.ts:
Changes here are actually coming from https://github.com/calcom/cal.com/pull/17510 which this branch is based on

sean-brydon commented on packages/ui/components/filter-select/index.tsx:
```diff
@@ -0,0 +1,102 @@
+  emptyText?: string;
+}
+
+export function FilterSelect({
```
Reusable style for filters based on what we used in data-table but not being tanstack table dependant

sean-brydon committed:
remove lock

sean-brydon committed:
Merge branch 'main' into feat/routingform-org-attribute-seed

sean-brydon committed:
Merge branch 'feat/routingform-org-attribute-seed' into feat/insights-routing-forms

sean-brydon committed:
Update packages/prisma/schema.prisma

joeauyeung commented on packages/features/insights/components/FailedBookingsByField.tsx:
```diff
@@ -0,0 +1,93 @@
+  return (
+    <div className="w-full text-sm">
+      <div className="flex h-12 items-center">
+        <h2 className="text-emphasis text-md font-semibold">Failed Bookings By Field</h2>
```
NIT: Missing translation

eunjae-lee replied:
I see this comment is resolved, but don't see the actual `t()` function here. Is the change not pushed yet?

joeauyeung commented on packages/features/insights/hooks/useDefaultRoutingForm.ts:
```diff
@@ -0,0 +1,60 @@
+  );
+
+  // Get the most popular form
+  const mostPopularForm = routingForms?.length
+    ? [...routingForms].sort((a, b) => (b._count?.responses ?? 0) - (a._count?.responses ?? 0))[0]
+    : null;
```
NIT: Could we do this server side in the `getRoutingFormsForFilters` query? Maybe sort the forms by number of responses?

sean-brydon replied:
We could be we already have the data here so it feels a bit redundant doing it in another endpoint when this just pulls from query cache

joeauyeung commented on packages/features/insights/server/routing-events.ts:
```diff
@@ -0,0 +1,459 @@
+      form: formsWhereCondition,
+    };
+
+    // Get total forms count
+    const totalForms = await prisma.app_RoutingForms_Form.count({
+      where: formsWhereCondition,
+    });
+
+    // Get total responses
+    const totalResponses = await prisma.app_RoutingForms_FormResponse.count({
+      where: responsesWhereCondition,
+    });
+
+    // Get responses without booking
+    const responsesWithoutBooking = await prisma.app_RoutingForms_FormResponse.count({
+      where: {
+        ...responsesWhereCondition,
+        routedToBookingUid: null,
+      },
+    });
```
I see three async calls that don't rely on each other. Can we wrap this in a `Promise.all()`?

joeauyeung commented on packages/features/insights/server/routing-events.ts:
```diff
@@ -0,0 +1,459 @@
+      .map(([formName, fields]) => ({
+        formName,
+        fields,
+        totalCount: Object.values(fields)
```
What is `totalCount`? I like the `reduce` that's used above but something about have a `reduce` inside of a `map` doesn't feel like it's the most efficient way of handling this. Could we instead calculate the `totalCount` in the first `reduce` that's called?

joeauyeung reviewed (changes requested):
Great work @sean-brydon. This was no easy task. 

The big thing I found was I wasn't able to stack filters on top of each other. I'd love to re-review this and test some more after this is fixed. I do have some comments right now though.

## Blocking
- It looks like adding more filters is replacing the current one
![image](https://github.com/user-attachments/assets/6af12327-744e-4604-8999-5e441b15fa17)
- When choosing to filter based on org it's not pulling any data. I don't think anyone would want to filter for the whole org so should we hide this option for now?
- It looks like filtering by event type isn't working as well. When I choose an event type it's still showing all routing form responses. I would expect it to only show responses that routed to that event type
- In the seeded form "Skills" field it looks like it's a multiselect but I can only choose one option at a time. Especially because that form field is a multi select.
- The downloaded CSV is coming up blank

## Non-blocking
- Should we be able to deselect an option? It feels like it should just default to "All"
![image](https://github.com/user-attachments/assets/0c622480-fe50-4903-8bbc-ddc2a43f73e2)
- Weird z index thing happening here
![image](https://github.com/user-attachments/assets/d800a1e5-0544-4767-b7ab-a09ccd8e8544)

sean-brydon committed:
remove event_type filter as it has no effect on routing forms

sean-brydon committed:
support multiple filters being rendered

sean-brydon commented:
> Great work @sean-brydon. This was no easy task.
> 
> The big thing I found was I wasn't able to stack filters on top of each other. I'd love to re-review this and test some more after this is fixed. I do have some comments right now though.
> 
> ## Blocking
> * It looks like adding more filters is replacing the current one
Fixed
![image](https://github.com/user-attachments/assets/add598f7-f45f-4539-b3df-14b8a2732985)


> * When choosing to filter based on org it's not pulling any data. I don't think anyone would want to filter for the whole org so should we hide this option for now?

> * It looks like filtering by event type isn't working as well. When I choose an event type it's still showing all routing form responses. I would expect it to only show responses that routed to that event type
Removed this option - you can't really filter routing forms by event type. 

> * In the seeded form "Skills" field it looks like it's a multiselect but I can only choose one option at a time. Especially because that form field is a multi select.

> * The downloaded CSV is coming up blank
 
> ## Non-blocking
> * Should we be able to deselect an option? It feels like it should just default to "All"
100% - this is a bug in insights also. Will fix.

sean-brydon committed:
omit org if in routing insights

sean-brydon committed:
wip download

sean-brydon committed:
download in batches

sean-brydon committed:
promise.all form counts

sean-brydon committed:
explain reduce and total count

sean-brydon committed:
type check

sean-brydon committed:
type check

sean-brydon committed:
more type fixes

sean-brydon committed:
attributes and download type fixes

sean-brydon committed:
ts server doesnt work for me

sean-brydon committed:
FC correct type

sean-brydon committed:
element[] to element ts error

sean-brydon committed:
Merge branch 'main' into feat/insights-routing-forms

sean-brydon committed:
Merge branch 'main' into feat/insights-routing-forms

sean-brydon committed:
restore yarn.lock

sean-brydon commented on packages/lib/csvUtils.ts:
```diff
@@ -21,6 +24,30 @@ export const downloadAsCsv = (csvRaw: string, filename: string) => {
   window.URL.revokeObjectURL(url);
 };
 
+export const objectsToCsv = (data: Record<string, any>[]): string => {
```
Object to csv instead of returning csv string from backend

eunjae-lee commented:
Massive work 🎉 I started reviewing this, but will continue Monday morning.

joeauyeung committed:
Fix zod parsing

joeauyeung committed:
Render selected routing form field filter

joeauyeung committed:
Add filtering for multiple routing form fields

joeauyeung commented on packages/features/insights/filters/index.tsx:
```diff
@@ -51,7 +61,9 @@ export const Filters = ({ showRoutingFilters = false }: { showRoutingFilters?: b
             <RoutingFormFilterList />
             <BookingStatusFilter />
-            <RoutingFormFieldFilter />
+            {routingFormFieldIds.map((fieldId) => {
```
I had to take this approach since I was getting a `Rendered more hooks than during the previous render` error trying to map the routing form field filters inside of `<RoutingFormFieldFilter />`

joeauyeung committed:
Fix test
